### PR TITLE
fix(files): file service now sends 304 and 403 headers more reliably

### LINKED ISSUE #9571: File service needs to delay eTag matching until after last updated time has been checked
We are sending `304` too early. Change in last updated time effectively means that URL should expire.

### PR DESCRIPTION
**Pull request by hypeJunction**
File service was sending 304 headers before testing if file modified time has changed.

Updates PHPUnit bootstrap to allow working with file entity objects. Modifies and adds new file related tests.

Adds two new methods to ElggFile: setModifiedTime and getModifiedTime, which invalidate stat cache
to ensure we always have latest modified time information

Fixes #9571

### CODE REVIEW

**Comment by hypeJunction** [comment]
This is a bug fix but also adds new API. Should we get it into 2.x or 2.1?

**Comment by hypeJunction** [comment]
In 2.1, we can just move the `eTag` logic down, and update tests and add API in 2.x

File: engine/classes/ElggFile.php [hunk 1]
```diff
@@ -318,6 +318,34 @@ public function tell() {
...
+	 * @return bool
+	 */
+	public function setModifiedTime($time = null, $atime = null) {
+		if (!isset($time)) {
+			$time = time();
+		}
+		if (!isset($atime)) {
+			$atime = $time;
+		}
+		if ($result = touch($this->getFilenameOnFilestore(), $time, $atime)) {
```
Comment (mrclay) [inline engine/classes/ElggFile.php]: Is setting particular time a really need? Seems like something that might be unreliable on some file systems.

**Comment by hypeJunction** [comment]
Breaking this down.

**Pull request closed without merging**